### PR TITLE
(dev/core/40) Show List of Contributions On Recurring Contribution View

### PR DESCRIPTION
**Pull request by MiyaNoctem**
Overview
----------------------------------------
Currently, when viewing a recurring contribution from contact's detailed view (on Contributions tab), the only information shown is its basic information. It would be great if we could also see the contributions related to that contribution on that same view.

Before
----------------------------------------
Clicking on a recurring contribution's **View** action only shows it's basic information on the opened modal dialog.

basw-79-before

After
----------------------------------------
Opening modal dialog to view a recurring contribution now show a table with a paginated list of associated contributions.

basw-79-after

Technical Details
----------------------------------------
Implemented by loading a new block via ajax

Comments
----------------------------------------
Issue created in gitlab issue tracker.
https://lab.civicrm.org/dev/core/issues/40

### CODE REVIEW

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
ping @eileenmcnaughton @mattwire @monishdeb @JoeMurray I quite like this idea and its a nice improvement IMO thoughts?

**Comment by eileenmcnaughton** [comment]
I think the UI improvement is good & it makes sense on view & edit views.

 In terms of implementation I'm less convinced. I think the way we have been going with the payments block & some other areas is to have a re-usable block that can show up in various places (often via ajax). This seems very much coded into a specific page.

File: CRM/Contribute/Page/ContributionRecur.php [hunk 1]
```diff
@@ -97,6 +101,30 @@ public function preProcess() {
...
+    );
+    $controller->setEmbedded(TRUE);
+    $controller->reset();
+    $controller->set('cid', $this->_contactId);
+    $controller->set('recur', $this->_id);
+    $controller->set('force', 1);
+    $controller->set('context', 'contribution');
+    $controller->set('limit', 50);
+    $controller->process();
+    $controller->run();
```
Comment (mattwire) [inline CRM/Contribute/Page/ContributionRecur.php]: Does this automatically create a pager if there are more than 50 contributions?

**Comment by mattwire** [comment]
I like it as a UI improvement.

@MiyaNoctem What does this do on an "Edit" recurring contribution view?
@eileenmcnaughton Can you point to the "payments block" ajax code?

I do think a reuseable block would be better if it was possible to do.  For example, you could then do things like rearrange the contribution tab/create a new recur contribution tab so contributions are grouped by recurring contribution.
Though I note that this PR is already reusing the search task so possibly most of the way there.

**Comment by eileenmcnaughton** [comment]
hmm - I think reusing search task puts us in the situation of re-using hard to use / read  / test/ maintain code :-( I think the search & sort functionality comes from the use of the datatables + ajax not the selector code per se. 

I'm trying to find a really good example. Somewhere between the way the transaction block shows when you expand the contribution   (civicrm/payment?transaction=view)

Although I hate the way it uses the same form for view as edit & just does something completely different on the same form!


```
  if ($this->_view == 'transaction' && ($this->_action & CRM_Core_Action::BROWSE)) {
      $title = $this->assignPaymentInfoBlock();
      CRM_Utils_System::setTitle($title);
      return;
    }
```

And the way the relationshipSelector stuff works - although I hate the face it doesn't have it's own class & just uses a function in the ajax class.

The activity block is also rendered via ajax (sadly the unit test was retrofitted onto it.

```
   <item>
     <path>civicrm/ajax/contactactivity</path>
     <page_callback>CRM_Activity_Page_AJAX::getContactActivity</page_callback>
     <access_arguments>access CiviCRM</access_arguments>
  </item>
```

**Comment by mattwire** [comment]
I wonder if there could be a halfway compromise, whereby the "search task" is loaded via an ajax function and could easily be replaced by improved/well-tested code in future?

**Comment by eileenmcnaughton** [comment]
probably best to get @colemanw input. 


I did get a fairly generic approach  mostly working in an extension - but some gaps there still from memory


https://github.com/eileenmcnaughton/org.wikimedia.omnimail/blob/master/CRM/Omnimail/Page/MailingsView.php
https://github.com/eileenmcnaughton/org.wikimedia.omnimail/blob/master/templates/CRM/Omnimail/Page/MailingsView.tpl

**Comment by JoeMurray** [comment]
I like the direction of the UI enhanvement but want it to be standardized in terms of design pattern with other places like payments when contributions on contact summary page contributions tab have the triangle clicked to open the view of them. I too think we should be trying to reuse implementation pattern of an includable file. I suppose we should concern ourselves with possibility of a pager when there are > 50 contributions, but I'd be willing to accept a poor job on that.

**Comment by MiyaNoctem** [comment]
@mattwire , on editing a recurring contribution, there would be no change. This is only shown on view action of a recurring contribution.

**Comment by MiyaNoctem** [comment]
@eileenmcnaughton, @mattwire,

I've re-written the how the related contributions are loaded and used something similar to what Eileen did for Omimil, basically load the contributions with the API, then process them for showing by adding the payment expansion control and actions (mong other fixes) and finally passing them to be dealt with by Smarty.

I've updated the screenshot to show how it works now.

@JoeMurray, you will also find the expanding triangle and the paging working.

Let me know what you guys think!

CC @colemanw

**Comment by eileenmcnaughton** [comment]
This looks good - I'm still a bit concerned in general about putting new visual blocks in their own class / tpl & my preference is that they be loaded by ajax  - but I defer to 'whatever @colemanw says' on this question.

Other than than that I agree this is an improvement

**Comment by guanhuan** [comment]
Thanks @eileenmcnaughton. @colemanw any thoughts?

**Comment by eileenmcnaughton** [comment]
Just a note  - I don't know if this is the best way but this is what I wound up using recently to load a visual block via ajax elsewhere

```
<div id="contact-summary-relationship-block" class="crm-clear crm-inline-block-content">

  <div class = 'crm-summary-row crm-label'>My special block</div>

</div>
<script>
  {literal}
  CRM.loadPage(
    CRM.url(
      'civicrm/relationshipblock/view',
      {"snippet" : 1, "contact_id" : {/literal}{$contactID}{literal}},
      'back'
    ),
    {
      target : '#contact-summary-relationship-block',
      dialog : false
    }
  );
{/literal}
</script>
```

**Comment by colemanw** [comment]
@eileenmcnaughton why ajax rather than a smarty include?

**Comment by eileenmcnaughton** [comment]
@colemanw if you say a smarty include is better you are the decider on that. My main concern is that it be re-usable & ideally should be a one-line to add it to a form php class & a one-line to add it to a tpl. It seems inevitable to me that this block will be re-used & wind up on both edit & view & should be written with that understanding

I do think bringing in by ajax can reduce initial page-load times - esp for things that are going to appear further down the page

**Comment by colemanw** [comment]
> I do think bringing in by ajax can reduce initial page-load times - esp for things that are going to appear further down the page

Interesting, my concern was the opposite - building a page by ajax can be quite slow because it involves bootstrapping the CMS and Civi once for every request. Were you thinking it would hamper the initial page load because this block involves slow queries?

**Comment by eileenmcnaughton** [comment]
@colemanw yeah - that was my perspective - there are a number of civi forms that are pretty slow to load because they do so much before they render.

**Comment by MiyaNoctem** [comment]
@eileenmcnaughton, @colemanw,

So the decision here is to do this using ajax calls? Can you please confirm?

**Comment by eileenmcnaughton** [comment]
@MiyaNoctem I would say @colemanw is on the fence so you don't have to go that way if you don't want to - but please move the php code out to it's own class & the tpl wrapper block to it's own file - so these can be included in other classes (or it could be exposed via ajax later if we wanted)

**Comment by MiyaNoctem** [comment]
@eileenmcnaughton, I've created a new page with it's own class and template to show the list of contributions and I'm loading it using an ajax call as you suggested. Let me know what you think!

**Comment by eileenmcnaughton** [comment]
Thanks @MiyaNoctem I'll take a look!

**Comment by eileenmcnaughton** [comment]
@MiyaNoctem I think the code looks good - it's really readable which is fantastic!

I hit one issue on testing - see the extraneous block of contribution summary chart stuff. From what I can tell this happens because you are using a sub-path of civicrm/contribute - which is what the dash is attached to - so it's loading that as well. I think that is a bit of a code issue & might merit some digging into at some point - but it would seem OK to me to just  change

<path>civicrm/contribute/contributionrecur-payments</path>

to 
<path>civicrm/contributionrecur-payments</path>

to bypass it

screenshot 2018-05-07 13 06 20

**Comment by eileenmcnaughton** [comment]
Actually I realise after reviewing the other PR that I was not getting the right view as I needed to do more cache-clearing. I now see

screenshot 2018-05-07 13 44 09

**Comment by eileenmcnaughton** [comment]
I'm going to squash and merge this - I generally prefer to get you to squash commits because it is a little better in the merge history. However, I found this conflicts with your other similar PR (the block on the membership view) and by getting this merged now I remove a block on that one.

**Pull request merged**